src/compute-client/src/controller.rs

```diff
@@ -236,6 +213,49 @@ impl<T> ComputeController<T> {
     }
 }
 
+impl<T> ComputeController<T>
+where
+    T: Timestamp + TotalOrder,
+{
+    /// Returns the write frontier for each installed collection.
+    pub fn write_frontiers(&self) -> BTreeMap<GlobalId, T> {
```

Comment:
> Why is this method returning a `T` and not an `Antichain<T>`?

Simply for historical reasons. `mz_cluster_replica_frontiers` existed before and took this approach of encoding the frontier as a single timestamp and omitting empty frontiers. I agree it's weird. Let me see how big a diff we get if I change this here, if it's too large I would rather open a follow-up PR.

> Also, the implementation filters out some installed collections, is that intended?

Do you mean the ones that have an empty frontier? Yes, that's intended (see above), and yeah the doc comment is lying. This would be resolved by returning an `Antichain` instead.